src/NextjsRevaluation.ts

```diff
@@ -0,0 +1,63 @@
+import { Duration, Stack } from 'aws-cdk-lib';
+import { Code, Function, FunctionOptions, Runtime } from 'aws-cdk-lib/aws-lambda';
+import { SqsEventSource } from 'aws-cdk-lib/aws-lambda-event-sources';
+import { Queue } from 'aws-cdk-lib/aws-sqs';
+import { Construct } from 'constructs';
+import { NextjsBaseProps } from './NextjsBase';
+import { NextjsBuild } from './NextjsBuild';
+import { NextJsLambda } from './NextjsLambda';
+
+export interface RevaluationProps extends NextjsBaseProps {
+  /**
+   * Override function properties.
+   */
+  readonly lambdaOptions?: FunctionOptions;
+
+  /**
+   * The `NextjsBuild` instance representing the built Nextjs application.
+   */
+  readonly nextBuild: NextjsBuild;
+
+  /**
+   * The main NextJS server handler lambda function.
+   */
+  readonly serverFunction: NextJsLambda;
+}
+
+/**
+ * Builds the system for revaluating Next.js resources. This includes a Lambda function handler and queue system.
+ */
+export class NextjsRevaluation extends Construct {
+  constructor(scope: Construct, id: string, props: RevaluationProps) {
+    super(scope, id);
+
+    if (!props.nextBuild) return;
+
+    const code = props.isPlaceholder
+      ? Code.fromInline(
+          "module.exports.handler = async () => { return { statusCode: 200, body: 'SST placeholder site' } }"
+        )
+      : Code.fromAsset(props.nextBuild.nextImageFnDir);
+
+    const queue = new Queue(this, 'RevalidationQueue', {
+      fifo: true,
+      receiveMessageWaitTime: Duration.seconds(20),
+    });
+    const consumer = new Function(this, 'RevalidationFunction', {
```

Comment:
This function will also need permissions to update cache s3 bucket.

Reply:
I'm not so sure about that actually (obviously please correct me though!) - [from what I can tell looking at open-next](https://github.com/serverless-stack/open-next/blob/274d446ed7e940cfbe7ce05a21108f4c854ee37a/packages/open-next/src/adapters/revalidate.ts#L18) it looks like the revalidation function just makes an HTTP request to the page. I don't know if I personally love this (vs the alternative of actually doing the work to generate the content required directly), but assuming this is indeed the case the revalidation function just needs access to the internet I think?

Reply:
Oh, worth mentioning that in my view open-next's architecture diagram isn't really clear about this, or is just flat out wrong? There should ben arrow pointing back to the server function or something maybe?

Reply:
@kevin-mitchell, I was wrong. Revalidation function should send HEAD requests to Static Assets bucket, not cache bucket so that Next.js re-creates them. open-next docs:
> Create a Lambda function using the code in the .open-next/revalidation-function folder, with the handler index.mjs.

Also, create an SQS FIFO queue, and set it as the event source for this function.

This function polls the queue for revalidation messages. Upon receiving a message, the function sends a HEAD request to the specified route for its revalidation.

However, a cache bucket does need to be created as noted in open-next docs:
> Create an S3 bucket and upload the content in the .open-next/cache folder to the root of the bucket. If you need to upload the files to a subfolder within the bucket, [refer to this section](https://github.com/serverless-stack/open-next/blob/main/README.md#reusing-same-bucket-for-asset-and-cache).

There are two types of caches in the .open-next/cache folder:

Route cache: This cache includes html and json files that are prerendered during the build. They are used to seed the revalidation cache.
Fetch cache: This cache includes fetch call responses, which might contain sensitive information. Make sure these files are not publicly accessible

Reply:
There is new a separate cache bucket.